src/libsystemd/sd-event/sd-event.c

```diff
@@ -1465,6 +1460,12 @@ _public_ int sd_event_add_child(
                 e->need_process_child = true;
         }
 
+        r = hashmap_put(e->child_sources, PID_TO_PTR(pid), s);
+        if (r < 0)
+                return r;
```

Comment:
hmm, if you swap this you need some clean-up patch changes here, i.e. undo the effect of source_child_pidfd_register on failure

Reply:
maybe the other option is to move the hashmap_put() further up, i.e. close to the old initialization of `s->child.pid`

Reply:
source_free → source_disconnect ↔ source-child_pidfd_unregister.
So this should be fine.